src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs

```diff
@@ -180,64 +181,67 @@ public struct AsyncTaskMethodBuilder<TResult>
             // result in a boxing allocation when storing the TStateMachine if it's a struct, but
             // this only happens in active debugging scenarios where such performance impact doesn't
             // matter.
-            if (taskField is AsyncStateMachineBox<IAsyncStateMachine> weaklyTypedBox)
+            else if (taskField is AsyncStateMachineBox<IAsyncStateMachine> weaklyTypedBox)
             {
                 // If this is the first await, we won't yet have a state machine, so store it.
```

Comment:
If we're trying to squeeze water from a stone, this whole block should be extremely cold, so if you can come up with a good way to outline it all to a separate helper, that'd be reasonable. Not sure if that'd help much, though, given that stateMachine is generic.

Reply:
Yeah, it would likely make things worse because it's still the same code, but with extra unwinding/GC tracking data structures because we have a new method.